util/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java

```diff
@@ -52,234 +48,99 @@
-    Set<EquivalentAddressGroup> removedAddrs = setsDifference(currentAddrs, latestAddrs.keySet());
-
-    for (Map.Entry<EquivalentAddressGroup, EquivalentAddressGroup> latestEntry :
-        latestAddrs.entrySet()) {
-      EquivalentAddressGroup strippedAddressGroup = latestEntry.getKey();
-      EquivalentAddressGroup originalAddressGroup = latestEntry.getValue();
-      Subchannel existingSubchannel = subchannels.get(strippedAddressGroup);
-      if (existingSubchannel != null) {
-        // EAG's Attributes may have changed.
-        existingSubchannel.updateAddresses(Collections.singletonList(originalAddressGroup));
-        continue;
-      }
-      // Create new subchannels for new addresses.
-
-      // NB(lukaszx0): we don't merge `attributes` with `subchannelAttr` because subchannel
-      // doesn't need them. They're describing the resolved server list but we're not taking
-      // any action based on this information.
-      Attributes.Builder subchannelAttrs = Attributes.newBuilder()
-          // NB(lukaszx0): because attributes are immutable we can't set new value for the key
-          // after creation but since we can mutate the values we leverage that and set
-          // AtomicReference which will allow mutating state info for given channel.
-          .set(STATE_INFO,
-              new Ref<>(ConnectivityStateInfo.forNonError(IDLE)));
-
-      final Subchannel subchannel = checkNotNull(
-          helper.createSubchannel(CreateSubchannelArgs.newBuilder()
-              .setAddresses(originalAddressGroup)
-              .setAttributes(subchannelAttrs.build())
-              .build()),
-          "subchannel");
-      subchannel.start(new SubchannelStateListener() {
-          @Override
-          public void onSubchannelState(ConnectivityStateInfo state) {
-            processSubchannelState(subchannel, state);
-          }
-        });
-      subchannels.put(strippedAddressGroup, subchannel);
-      subchannel.requestConnection();
-    }
-
-    ArrayList<Subchannel> removedSubchannels = new ArrayList<>();
-    for (EquivalentAddressGroup addressGroup : removedAddrs) {
-      removedSubchannels.add(subchannels.remove(addressGroup));
-    }
-
-    // Update the picker before shutting down the subchannels, to reduce the chance of the race
-    // between picking a subchannel and shutting it down.
-    updateBalancingState();
-
-    // Shutdown removed subchannels
-    for (Subchannel removedSubchannel : removedSubchannels) {
-      shutdownSubchannel(removedSubchannel);
-    }
-
-    return true;
+  protected SubchannelPicker getSubchannelPicker(Map<Object, SubchannelPicker> childPickers) {
+    return null; // TODO
   }
 
   @Override
-  public void handleNameResolutionError(Status error) {
-    if (currentState != READY)  {
-      updateBalancingState(TRANSIENT_FAILURE, new EmptyPicker(error));
-    }
-  }
+  protected ResolvedAddresses getChildAddresses(Object key, ResolvedAddresses resolvedAddresses,
```

Comment:
So `MultiChildLoadBalancer` calls the abstract `getChildAddresses()` which is implemented here, but then this just uses the return value from `pickResolvedAddress()` that is implemented in `MultiChildLoadBalancer`. Is this bouncing around needed?

I noticed that `ClusterManagerLoadBalancer` also overrides this for the purpose of modifying the LB config. Would it make this simpler to have `MultiChildLoadBalancer` just use what `pickResolvedAddresses()` provides and then have an optional facility to have an extending LB modify the `ResolvedAddresses` if it needs to, instead of forcing `getChildAddresses()` to be implemented?

Reply:
That's a good suggestion.  A bunch of the other methods went from abstract to defaults, this one can also.